taichi/transforms/demote_atomics.cpp

```diff
@@ -41,42 +42,55 @@ class DemoteAtomics : public BasicStmtVisitor {
+              demote = false;
+              break;
+            }
+            if (current_offloaded->mem_access_opt.has_flag(
+                    snode, SNodeAccessFlag::block_local) ||
+                current_offloaded->mem_access_opt.has_flag(
+                    snode, SNodeAccessFlag::mesh_local)) {
+              // BLS does not support write access yet so we keep atomic_adds.
+              demote = false;
+              break;
+            }
+          }
+          // demote from-end atomics
+          if (current_offloaded->task_type == OffloadedTaskType::mesh_for) {
+            if (dest->indices.size() == 1 &&
+                dest->indices[0]->is<MeshIndexConversionStmt>()) {
+              auto idx = dest->indices[0]->as<MeshIndexConversionStmt>()->idx;
+              while (idx->is<MeshIndexConversionStmt>()) {  // special case: l2g
+                                                            // + g2r
+                idx = idx->as<MeshIndexConversionStmt>()->idx;
+              }
+              if (idx->is<LoopIndexStmt>() &&
+                  idx->as<LoopIndexStmt>()->is_mesh_index() &&
+                  loop_unique_ptr_[stmt->dest->as<GlobalPtrStmt>()
+                                       ->snodes.data[0]] != nullptr) {
+                demote = true;
+              }
+            }
           }
-          if (current_offloaded->mem_access_opt.has_flag(
-                  snode, SNodeAccessFlag::block_local) ||
-              current_offloaded->mem_access_opt.has_flag(
-                  snode, SNodeAccessFlag::mesh_local)) {
-            // BLS does not support write access yet so we keep atomic_adds.
+        } else if (stmt->dest->is<ExternalPtrStmt>()) {
+          ExternalPtrStmt *dest_ptr = stmt->dest->as<ExternalPtrStmt>();
+          demote = true;
+          if (loop_unique_arr_ptr_.find(dest_ptr) ==
+              loop_unique_arr_ptr_.end()) {
+            // Not loop unique
             demote = false;
-            break;
           }
-        }
-        // demote from-end atomics
-        if (current_offloaded->task_type == OffloadedTaskType::mesh_for) {
-          if (dest->indices.size() == 1 &&
-              dest->indices[0]->is<MeshIndexConversionStmt>()) {
-            auto idx = dest->indices[0]->as<MeshIndexConversionStmt>()->idx;
-            while (idx->is<MeshIndexConversionStmt>()) {  // special case: l2g +
-                                                          // g2r
-              idx = idx->as<MeshIndexConversionStmt>()->idx;
-            }
-            if (idx->is<LoopIndexStmt>() &&
-                idx->as<LoopIndexStmt>()->is_mesh_index() &&
-                loop_unique_ptr_[stmt->dest->as<GlobalPtrStmt>()
-                                     ->snodes.data[0]] != nullptr) {
-              demote = true;
-            }
+          if (dest_ptr->indices.empty()) {
+            demote = false;
           }
+          // TODO: Is BLS / Mem Access Opt a thing for any_arr?
```

Comment:
Good question, we might be able to extend BLS to external arrays as well.  However, at this point BLS depends on knowing the struct-for's loop index